File: nagasenacs/src/Nagasena/Proc/IO/Compression/EXIDeflaterOutputStream.cs
using System;
using System.IO;

using ICSharpCode.SharpZipLib.Zip.Compression;

namespace Nagasena.Proc.IO.Compression {

  internal sealed class EXIDeflaterOutputStream : Stream {

    private readonly Stream m_outputStream;

    private readonly Deflater m_deflater;

    private readonly byte[] m_bytesIn;
    private int m_n_bytesIn;

    private readonly byte[] m_bytesOut;

    public override bool CanRead {
      get {
        return false;
      }
    }
    public override bool CanSeek {
      get {
        return false;
      }
    }
    public override bool CanWrite {
      get {
        return true;
      }
    }
    public override long Length {
      get {
        throw new NotSupportedException();
      }
    }
    public override long Position {
      get {
        throw new NotSupportedException();
      }
      set {
        throw new NotSupportedException();
      }
    }

    public EXIDeflaterOutputStream(Stream outputStream, Deflater deflater) {
      m_outputStream = outputStream;
      deflater.Reset();
      m_deflater = deflater;

      m_bytesIn = new byte[65536];
      m_n_bytesIn = 0;
      m_bytesOut = new byte[8192];
    }

    public override void Flush() {
      flushInput();
      m_outputStream.Flush();
    }

    public override void Close() {
      m_outputStream.Close();
    }

    public override void WriteByte(byte bt) {
      m_bytesIn[m_n_bytesIn] = (byte)bt;
      if (++m_n_bytesIn == m_bytesIn.Length) {
        flushInput();
      }
    }

    private void flushInput() {
      long initialBytesRead = m_deflater.TotalIn;
      m_deflater.SetInput(m_bytesIn, 0, m_n_bytesIn);
      int n_bytesOut;
      do {
        if ((n_bytesOut = m_deflater.Deflate(m_bytesOut)) != 0) {
          for (int i = 0; i < n_bytesOut; i++) {
            m_outputStream.WriteByte((byte)(0xFF & m_bytesOut[i]));
          }
        }
        else {
          break;
        }
      }
      while (true);
      int bytesRead = (int)(m_deflater.TotalIn - initialBytesRead);
      if (bytesRead == m_n_bytesIn) {
        m_n_bytesIn = 0;
      }
      else if (bytesRead != 0) {
        m_n_bytesIn -= bytesRead;
        Array.Copy(m_bytesIn, bytesRead, m_bytesIn, 0, m_n_bytesIn);
      }
    }

    internal void resetDeflater() {
      if (m_n_bytesIn != 0) {
        flushInput();
      }
      m_deflater.SetInput(m_bytesIn, 0, m_n_bytesIn);
      m_deflater.Finish();
      while (!m_deflater.IsFinished) {
        int len = m_deflater.Deflate(m_bytesOut);
        for (int i = 0; i < len; i++) {
          m_outputStream.WriteByte((byte)(0xFF & m_bytesOut[i]));
        }
      }
      m_deflater.Reset();
    }

    ///////////////////////////////////////////////////////////////////////////
    /// Stream Functions
    ///////////////////////////////////////////////////////////////////////////

    public override long Seek(long offset, SeekOrigin origin) {
      throw new NotSupportedException();
    }

    public override void SetLength(long value) {
      throw new NotSupportedException();
    }

    public override int Read(byte[] buffer, int offset, int count) {
      throw new NotSupportedException();
    }

    public override void Write(byte[] buffer, int offset, int count) {
      throw new NotSupportedException();
    }

  }

}

File: sumerogi/src/com/sumerogi/proc/io/compression/EXIDeflaterOutputStream.java
package com.sumerogi.proc.io.compression;

import java.io.IOException;
import java.io.OutputStream;

import java.util.zip.Deflater;

final class EXIDeflaterOutputStream extends OutputStream {
  
  private final OutputStream m_outputStream;

  private final Deflater m_deflater;
  
  private final byte[] m_bytesIn;
  private int m_n_bytesIn;
  
  private final byte[] m_bytesOut;
  
  public EXIDeflaterOutputStream(OutputStream outputStream, Deflater deflater) {
    m_outputStream = outputStream;
    deflater.reset();
    m_deflater = deflater; 
    
    m_bytesIn  = new byte[65536];
    m_n_bytesIn = 0;    
    m_bytesOut = new byte[8192];
  }

  @Override
  public void flush() throws IOException {
    flushInput();
    m_outputStream.flush();
  }
  
  @Override
  public void close() throws IOException {
    m_outputStream.close();
  }
  
  @Override
  public void write(int bt) throws IOException {
    m_bytesIn[m_n_bytesIn] = (byte)bt;
    if (++m_n_bytesIn == m_bytesIn.length) {
      flushInput();
    }
  }
  
  private void flushInput() throws IOException {
    final long initialBytesRead = m_deflater.getBytesRead();
    m_deflater.setInput(m_bytesIn, 0, m_n_bytesIn);
    int n_bytesOut;
    do {
      if ((n_bytesOut = m_deflater.deflate(m_bytesOut)) != 0) { 
        for (int i = 0; i < n_bytesOut; i++) {
          m_outputStream.write(0xFF & m_bytesOut[i]);
        }
      }
      else
        break;
    }
    while (true);
    final int bytesRead = (int)(m_deflater.getBytesRead() - initialBytesRead);
    if (bytesRead == m_n_bytesIn)
      m_n_bytesIn = 0;
    else if (bytesRead != 0) {
      m_n_bytesIn -= bytesRead;
      System.arraycopy(m_bytesIn, bytesRead, m_bytesIn, 0, m_n_bytesIn);
    }
  }

  void resetDeflater() throws IOException {
    if (m_n_bytesIn != 0)
      flushInput();
    m_deflater.setInput(m_bytesIn, 0, m_n_bytesIn);
    m_deflater.finish();
    while (!m_deflater.finished()) {
      int len = m_deflater.deflate(m_bytesOut);
      for (int i = 0; i < len; i++) {
        m_outputStream.write(0xFF & m_bytesOut[i]);
      }
    }
    m_deflater.reset();
  }
  
}
